File: BOJ/랜덤디펜스/1890.js
// 정답 범위가 최대 2^63-1이므로 BigInt를 사용
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")
  .map((x) => x.split(" ").map((x) => +x));

const n = input[0][0];
const a = input.slice(1);
const d = Array(n)
  .fill()
  .map(() => Array(n).fill(0n));
d[0][0] = 1n;
for (let i = 0; i < n; i++) {
  for (let j = 0; j < n; j++) {
    const step = a[i][j];
    if (step === 0) {
      continue;
    }
    if (i + step < n) {
      d[i + step][j] += d[i][j];
    }
    if (j + step < n) {
      d[i][j + step] += d[i][j];
    }
  }
}
console.log(d[n - 1][n - 1].toString());

File: BOJ/랜덤디펜스/1890.py
import sys
input = sys.stdin.readline

n = int(input())
a = [list(map(int, input().split())) for _ in range(n)]
d = [[0] * n for _ in range(n)]
d[0][0] = 1
for i in range(n):
  for j in range(n):
    step = a[i][j]
    if step == 0:
      continue
    if i + step < n:
      d[i + step][j] += d[i][j]
    if j + step < n:
      d[i][j + step] += d[i][j]
print(d[-1][-1])